library/ssl_tls13_generic.c

```diff
@@ -1818,23 +1720,58 @@ static int ssl_certificate_verify_coordinate( mbedtls_ssl_context* ssl )
         return( MBEDTLS_ERR_SSL_BAD_HS_CERTIFICATE_VERIFY );
     }
 
-    /*
-     * Check whether the signature scheme corresponds to the hash algorithm of the negotiated ciphersuite
-     * TBD: Double-check whether this is really a good approach.
-
-     if( ( ssl->handshake->signature_scheme == SIGNATURE_ECDSA_SECP256r1_SHA256 ) && ( ciphersuite_info->hash != MBEDTLS_MD_SHA256 ) ) {
-     MBEDTLS_SSL_DEBUG_MSG( 1, ( "Certificate Verify: SIGNATURE_ECDSA_SECP256r1_SHA256 only matches with MBEDTLS_MD_SHA256." ) );
-     return( MBEDTLS_ERR_SSL_BAD_HS_CERTIFICATE_VERIFY );
-     }
-     else if( ( ssl->handshake->signature_scheme == SIGNATURE_ECDSA_SECP384r1_SHA384 ) && ( ciphersuite_info->hash != MBEDTLS_MD_SHA384 ) ) {
-     MBEDTLS_SSL_DEBUG_MSG( 1, ( "Certificate Verify: SIGNATURE_ECDSA_SECP384r1_SHA384 only matches with MBEDTLS_MD_SHA384." ) );
-     return( MBEDTLS_ERR_SSL_BAD_HS_CERTIFICATE_VERIFY );
-     }
-     else if( ( ssl->handshake->signature_scheme == SIGNATURE_ECDSA_SECP521r1_SHA512 ) && ( ciphersuite_info->hash != MBEDTLS_MD_SHA512 ) ) {
-     MBEDTLS_SSL_DEBUG_MSG( 1, ( "Certificate Verify: SIGNATURE_ECDSA_SECP521r1_SHA512 only matches with MBEDTLS_MD_SHA512." ) );
-     return( MBEDTLS_ERR_SSL_BAD_HS_CERTIFICATE_VERIFY );
-     }
-    */
+    /* Calculate the transcript hash */
+#if defined(MBEDTLS_SHA256_C)
+    if( ssl->handshake->ciphersuite_info->mac == MBEDTLS_MD_SHA256 )
+    {
+        mbedtls_sha256_init( &sha256 );
+
+        mbedtls_sha256_clone( &sha256, &ssl->handshake->fin_sha256 );
+
+        if( ( ret = mbedtls_sha256_finish_ret( &sha256, ssl->handshake->state_local.certificate_verify_out.handshake_hash ) ) != 0 )
```

Comment:
Minor: Overly long line.

Reply:
I shortened it but I am not sure it is much more readable now.